From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mio <mio@mio19.uk>
Date: Tue, 20 Oct 2026 19:13:29 +1100
Subject: [PATCH] overlay/asteroidos: port 2048 to Qt 6

Apply the twenty48 Qt 6 compatibility patch to MagneFire/asteroid-2048 so
the sliding-tile puzzle builds against the current AsteroidOS Qt 6
toolkit.

Signed-off-by: Mio <mio@mio19.uk>
---
 overlay/asteroidos/apps/default.nix        |   3 +
 overlay/asteroidos/apps/twenty48-qt6.patch | 124 +++++++++++++++++++++
 2 files changed, 127 insertions(+)
 create mode 100644 overlay/asteroidos/apps/twenty48-qt6.patch

diff --git a/overlay/asteroidos/apps/default.nix b/overlay/asteroidos/apps/default.nix
index df130737..4c56e6d1 100644
--- a/overlay/asteroidos/apps/default.nix
+++ b/overlay/asteroidos/apps/default.nix
@@ -136,5 +136,8 @@
     desc = "Sliding-tile 2048 puzzle game for AsteroidOS smartwatches";
     rev = "99e7f00228c269069b41c4a27a0a60d2ed2ac7cf";
     sha256 = "16xzqx0ry4y8hx6gqigafqx6izjahjlz5jc5s12fg0rjk5ylxjx0";
+    patches = [
+      ./twenty48-qt6.patch
+    ];
   };
 }
diff --git a/overlay/asteroidos/apps/twenty48-qt6.patch b/overlay/asteroidos/apps/twenty48-qt6.patch
new file mode 100644
index 00000000..f9607f14
--- /dev/null
+++ b/overlay/asteroidos/apps/twenty48-qt6.patch
@@ -0,0 +1,124 @@
+diff -Naur a/asteroid-2048.desktop.template b/asteroid-2048.desktop.template
+--- a/asteroid-2048.desktop.template	2026-10-18 16:11:39.862790321 +1100
++++ b/asteroid-2048.desktop.template	2026-10-18 16:11:50.761792981 +1100
+@@ -1,7 +1,8 @@
+ [Desktop Entry]
+ Type=Application
+ Categories=Applications;
+-Exec=invoker --single-instance --type=qtcomponents-qt5 /usr/bin/asteroid-2048
++Exec=asteroid-2048
+ Icon=2048
+-X-Asteroid-Center-Color=#b04d1c
+-X-Asteroid-Outer-Color=#421c0a
++Name=2048
++X-Asteroid-Center-Color=#4b45b9
++X-Asteroid-Outer-Color=#161537
+diff -Naur a/CMakeLists.txt b/CMakeLists.txt
+--- a/CMakeLists.txt	2026-10-18 16:11:39.862790321 +1100
++++ b/CMakeLists.txt	2026-10-18 16:11:50.753792979 +1100
+@@ -1,22 +1,33 @@
+-cmake_minimum_required(VERSION 3.6.0)
++cmake_minimum_required(VERSION 3.20)
+ 
+ project(asteroid-2048
+-	VERSION 0.0.1
++	VERSION 1.0.0
+ 	DESCRIPTION "A 2048 game clone for AsteroidOS")
+ 
+ find_package(ECM REQUIRED NO_MODULE)
+-find_package(AsteroidApp REQUIRED)
++find_package(AsteroidApp6 REQUIRED)
++find_package(Qt6 REQUIRED COMPONENTS Quick)
+ 
+ set(CMAKE_MODULE_PATH ${ECM_MODULE_PATH} ${ASTEROID_MODULE_PATH})
+ 
+ include(FeatureSummary)
+ include(GNUInstallDirs)
+-include(ECMQMLModules)
+-include(AsteroidCMakeSettings)
+-include(AsteroidTranslations)
++include(Asteroid6CMakeSettings)
++include(Asteroid6Translations)
+ 
+ add_subdirectory(src)
+ 
++configure_file(
++	${CMAKE_CURRENT_SOURCE_DIR}/runscript.in
++	${CMAKE_BINARY_DIR}/${CMAKE_PROJECT_NAME}
++	@ONLY
++)
++
++install(
++	PROGRAMS ${CMAKE_BINARY_DIR}/${CMAKE_PROJECT_NAME}
++	DESTINATION ${CMAKE_INSTALL_BINDIR}
++)
++
+ build_translations(i18n)
+ generate_desktop(${CMAKE_SOURCE_DIR} asteroid-2048)
+ 
+diff -Naur a/runscript.in b/runscript.in
+--- a/runscript.in	1970-01-01 10:00:00.000000000 +1000
++++ b/runscript.in	2026-10-18 16:11:50.755792979 +1100
+@@ -0,0 +1,3 @@
++#!/bin/sh
++
++exec invoker --single-instance --type=asteroid-qt6 @CMAKE_INSTALL_FULL_LIBDIR@/asteroid-2048.so
+diff -Naur a/src/CMakeLists.txt b/src/CMakeLists.txt
+--- a/src/CMakeLists.txt	2026-10-18 16:11:39.876790324 +1100
++++ b/src/CMakeLists.txt	2026-10-18 16:11:50.757792980 +1100
+@@ -1,8 +1,24 @@
+-add_library(asteroid-2048 main.cpp resources.qrc)
+-set_target_properties(asteroid-2048 PROPERTIES PREFIX "" SUFFIX "")
++add_library(${CMAKE_PROJECT_NAME}
++	main.cpp
++)
++set_target_properties(${CMAKE_PROJECT_NAME} PROPERTIES PREFIX "")
+ 
+-target_link_libraries(asteroid-2048 PUBLIC
+-	AsteroidApp)
++qt6_add_qml_module(${CMAKE_PROJECT_NAME}
++	URI Twenty48
++	VERSION 1.0
++	RESOURCE_PREFIX "/"
++	NO_PLUGIN
++	QML_FILES
++		main.qml
++		Game.qml
++)
+ 
+-install(TARGETS asteroid-2048
+-	DESTINATION ${CMAKE_INSTALL_BINDIR})
+\ No newline at end of file
++target_link_libraries(${CMAKE_PROJECT_NAME}
++	PUBLIC
++	Asteroid::AsteroidApp6
++	Qt6::Quick
++)
++
++install(TARGETS ${CMAKE_PROJECT_NAME}
++	DESTINATION ${CMAKE_INSTALL_LIBDIR}
++)
+diff -Naur a/src/main.cpp b/src/main.cpp
+--- a/src/main.cpp	2026-10-18 16:11:39.875790324 +1100
++++ b/src/main.cpp	2026-10-18 16:11:50.759792980 +1100
+@@ -16,8 +16,19 @@
+  */
+ 
+ #include <asteroidapp.h>
++#include <QGuiApplication>
++#include <QQmlContext>
++#include <QQuickView>
++#include <QScopedPointer>
++#include <QScreen>
++#include <QUrl>
+ 
+ int main(int argc, char *argv[])
+ {
+-    return AsteroidApp::main(argc, argv);
++    QScopedPointer<QGuiApplication> app(AsteroidApp::application(argc, argv));
++    QScopedPointer<QQuickView> view(AsteroidApp::createView());
++    view->setSource(QUrl(QStringLiteral("qrc:/Twenty48/main.qml")));
++    view->resize(app->primaryScreen()->size());
++    view->show();
++    return app->exec();
+ }
